### DISCUSSION TITLE
NPCs don't appear

### DISCUSSION THREAD

**Discussion by para-dise** [Minestom/Minestom#828] [Help] [answered]
So I tried the code from the wiki to spawn a horse
```java
Instance instance = ...;
Position spawnPosition = new Pos(0D, 42D, 0D);
Entity horse = new Entity(EntityType.HORSE);
horse.setInstance(instance, spawnPosition);
```

and this worked perfectly, but then I tried to spawn an NPC with

```java
Pos spawnPosition = new Pos(-4.779, 68.0, 126.5);
EntityCreature player = new EntityCreature(EntityType.PLAYER);
player.setUuid(UUID.fromString("069a79f4-44e9-4726-a5be-fca90e38aaf5"));
player.setInstance(this.instance, spawnPosition);
player.setAutoViewable(true);

handler.addListener(PlayerSpawnEvent.class, event -> {
    player.addViewer(event.getPlayer());
});
```

Am I missing something here?

EDIT:

As a side note, the player seems to actually collide with them, but they're still invisible

**Comment by collinistevens** [comment, marked as answer]
When creating a NPC via a living entity you need to send the proper packets for it to appear.

For actually showing the NPC you need to first spawn it and then send a `PlayerInfoPacket`

I've created this method for easily creating the packet:
```java
public static PlayerInfoPacket addPlayerInfoPacket(UUID uuid, String username, PlayerSkin skin) {
        var textureProperty = new PlayerInfoPacket.AddPlayer.Property("textures", skin.textures(), skin.signature());
        var playerEntry = new PlayerInfoPacket.AddPlayer(uuid, username, Collections.singletonList(textureProperty), GameMode.CREATIVE, 0, null);

        return new PlayerInfoPacket(PlayerInfoPacket.Action.ADD_PLAYER, Collections.singletonList(playerEntry));
    }
```

This packet will add the player to the tab list and actually make it display client side, you should be sending this packet when the player comes into view with the NPC.


**Note:** If you have an NPC within render distance when the player logs in, you should send that packet with a few seconds of delay as adding and removing instantly can cause the skin to not show properly.

If you then want to remove it from the tab list you need to send another `PlayerInfoPacket` with the remove player action.

```java
public static PlayerInfoPacket removePlayerInfoPacket(UUID uuid) {
        var playerEntry = new PlayerInfoPacket.RemovePlayer(uuid);

        return new PlayerInfoPacket(PlayerInfoPacket.Action.REMOVE_PLAYER, Collections.singletonList(playerEntry));
    }
```

**Reply by RinesThaix** [reply]
A note is not precisely correct. Minecraft client works in a way that even if you have a player info packet received from the server, skin won't be loaded until NPC is visible on your screen. So, you can delay tablist removal even for a thousand years, but if a player won't rotate to the direction of NPC or there will be any blocks preventing player from seeing an NPC, skin won't load.
There are several solutions to resolve that issue, but the most appropriate one is to spawn NPC inside player whenever he enters NPC's vision range and then, one tick later, teleport NPC to it's real position. Please be aware that I'm not talking about true teleportation: it's just a trick with packets. In terms on Minestom it can be achieved by overriding spawn packet instantiation or by sending two teleportation events: one when `Entity#updateNewViewer(Player)` is called and another -- 1 tick later.

**Reply by amrouda50** [reply]
> When creating a NPC via a living entity you need to send the proper packets for it to appear.
> 
> For actually showing the NPC you need to first spawn it and then send a `PlayerInfoPacket`
> 
> I've created this method for easily creating the packet:
> 
> ```java
> public static PlayerInfoPacket addPlayerInfoPacket(UUID uuid, String username, PlayerSkin skin) {
>         var textureProperty = new PlayerInfoPacket.AddPlayer.Property("textures", skin.textures(), skin.signature());
>         var playerEntry = new PlayerInfoPacket.AddPlayer(uuid, username, Collections.singletonList(textureProperty), GameMode.CREATIVE, 0, null);
> 
>         return new PlayerInfoPacket(PlayerInfoPacket.Action.ADD_PLAYER, Collections.singletonList(playerEntry));
>     }
> ```
> 
> This packet will add the player to the tab list and actually make it display client side, you should be sending this packet when the player comes into view with the NPC.
> 
> **Note:** If you have an NPC within render distance when the player logs in, you should send that packet with a few seconds of delay as adding and removing instantly can cause the skin to not show properly.
> 
> If you then want to remove it from the tab list you need to send another `PlayerInfoPacket` with the remove player action.
> 
> ```java
> public static PlayerInfoPacket removePlayerInfoPacket(UUID uuid) {
>         var playerEntry = new PlayerInfoPacket.RemovePlayer(uuid);
> 
>         return new PlayerInfoPacket(PlayerInfoPacket.Action.REMOVE_PLAYER, Collections.singletonList(playerEntry));
>     }
> ```

and how do u actually send this packet after creating it by the function?
